### ISSUE TITLE
Re-architect unit testing (automate different platforms, mostly)

### ISSUE THREAD

**Issue by wlandau-lilly** [wlandau-lilly/drake#58]
Drake needs to work for all environments, platforms, modes of parallelism, and levels of scale. Some of these settings are controlled at the top of [debug.R](https://github.com/wlandau-lilly/drake/blob/master/R/debug.R) and [test.R](https://github.com/wlandau-lilly/drake/blob/master/R/test.R), which are then passed to `testrun()` for most of the [unit tests](https://github.com/wlandau-lilly/drake/tree/master/tests/testthat). For CRAN, the settings are:

- `envir = new.env(parent = globalenv())`
- `parallelism = "parLapply"`
- `jobs = 2`

I have to manually change [debug.R](https://github.com/wlandau-lilly/drake/blob/master/R/debug.R) and [test.R](https://github.com/wlandau-lilly/drake/blob/master/R/test.R) every time I want to change these settings and retest, and the combinatorics are killing me. It would really help to have an automated way to run through the following table of settings and still leave the CRAN settings alone at a single configuration (second row of the table). 

| envir | parallelism | jobs | notes |
|---|---|---|---|
| `new.env(parent = globalenv())` | `"parLapply"` | 1 | Uses `lapply()` instead of `parLapply()` when `jobs = 1`. | 
| `new.env(parent = globalenv())` | `"parLapply"` | 2 | For CRAN, Travis, and Appveyor, only use this configuration. |
| `new.env(parent = globalenv())` | `"mclapply"` | 1 | 
| `new.env(parent = globalenv())` | `"mclapply"` | 8 | Skip on Windows. |
| `new.env(parent = globalenv())` | `"Makefile"` | 1 | |
| `new.env(parent = globalenv())` | `"Makefile"` | 2 | Makefiles are different, so I want to test with a low and a high `jobs` value. |
| `new.env(parent = globalenv())` | `"Makefile"` | 16 | 
| `globalenv()` | `"parLapply"` | 1 | For the global environment, I do not think all scenarios need to be repeated. | 
| `globalenv()` | `"parLapply"` | 2 |
| `new.env()` | `"mclapply"` | 1 | 
| `globalenv()` | `"mclapply"` | 8 | Skip on Windows. |
| `globalenv()` | `"Makefile"` | 16 |

@krlmlr's [mockr](https://github.com/krlmlr/mockr) might help here, but I am not sure.

**Comment by wlandau-lilly** [comment]
@AlexAxthelm's #59 was helpful for thought, and here are some more ideas.

Currently, here are the tests in `tests/testthat/` that should be tested for all configuration settings in the table.
- `test-basic.R`
- `test-command-changes.R`
- `test-envir.R`
- `test-full-build.R`
- `test-import-file.R`
- `test-import-object.R`
- `test-intermediate-file.R`
- `test-reproducible.R` (after I merge #56)

To accomplish this, I currently use a clumsy combination of `dbug()$envir` (top of `R/debug.R`), `testopts()`, `testrun()`, and `testrun_automatic_packages()` (all in `R/test.R`). I propose an alternative framework.

1. Dispense with the existing `testopts()`.
2. Borrow the `test_configs()` function from #59. Rename it `all_test_opts()` to avoid confusion with the `config` list in the internals of drake.
3.  Define `test_opt()`, which picks a setting from `all_test_opts()`.

```r
test_opt <- function(opt = "parent_parL_2"){
   all_test_opts()[[opt]]
}
```

4. Use only the default `test_opt()` in `tests/testthat.R`
5. Define a new `tests/testthat_exhaustive.R` to loop over all testing scenarios. I am picturing something like the following, but I do not know if it will actually work.

```r
Sys.setenv("R_TESTS" = "")
library(testthat)
library(drake)
for(opt in names(all_test_opts()){
  if(testing_scenario_should_be_skipped(opt)) # Decide based on the OS
    next
  formals(test_opt) <- list(opt = opt)
  test_check("drake")
}
```

In addition, the testing code needs to be cleaned up.

1. Expose new functions `dbug_envir()`, `dbug_plan()`, and `dbug_files()` to be called within `dbug()`. Use these functions to replace clumsy uses of `dbug()` where needed.

```r
dbug_envir <- function(envir){
  force(envir)
  imports <- c(
    "f <- function(x) {g(x) + a}",
    "g <- function(y) {h(y) + b}",
    "h <- function(y) {i(y) + j(y)}",
    "i <- function(x) {x+1}",
    "j <- function(x) {x+2 + c}",
    "a <- 15",
    "b <- 20",
    "c <- 25")
  for(import in imports)
    eval(parse(text = import), envir = envir)
}

dbug_plan <- function(){
  plan(list = c(
    "'intermediatefile.rds'" = 
      "saveRDS(combined, \"intermediatefile.rds\")",
    yourinput = "f(1+1)",
    nextone = "myinput + g(7)",
    combined = "nextone + yourinput",
    myinput = "readRDS('input.rds')",
    final = "readRDS('intermediatefile.rds')"
  ))
}

dbug_files <- function(){
  saveRDS(1:10, "input.rds")
}
```

2. Connect `dbug()` to `test_opt()`.

```r
dbug <- function(){
  opts <- test_opt()
  envir <- eval(parse(text = opts$envir)) %>% dbug_envir
  dbug_files()
  plan <- dbug_plan()
  build_config(
    plan = plan,
     targets = plan$target
    parallelism = opts$parallelism,
    jobs = opts$jobs,
    ...
}
```

3. Modify `testrun()` to just react to the `config` argument, as opposed to a combination of `config` and `testopts()`.
4. Dispense with `testrun_automatic_packages()` because it is called only once. Instead, just use the appropriate call to `make()` in `test-Makefile.R`.

I may work on all these changes myself, but I am in a funny situation where I can spell things out in the issue tracker, but I need to go though a long disclosure process to modify the code base myself. I am working to fix company policy on that. In the meantime, those of you who are external can freely submit pull requests and run circles around me.

**Comment by AlexAxthelm** [comment]
regarding working directories: it would probably be safer to use `withr::with_dir()`

**Comment by wlandau-lilly** [comment]
Absolutely. Come to think of it, I am almost surprised `testthat` does not have native functionality for this.

**Comment by AlexAxthelm** [comment]
I think the expected behavior for a lot of folks using `testthat` is to let Travis handle the heavy lifting.

which leads me to a thought that we could have the tests check to see if there is an environmental variable set for the parallelism and jobs variables (`$R_DRAKE_PARALLELISM` and `$R_DRAKE_JOBS`, perhaps), and use a [build matrix on Travis](https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix), to actually test multiple sets of options. The default, if the variables are not set, would be the CRAN options.

 Locally, we could switch options using `withr::with_envvar()`:
```r
with_envvar(
  c(
    "R_DRAKE_PARALLELISM" = "Makefile",
    "R_DRAKE_JOBS" = 16
    ),
  devtools::test()
)
```
This would make local testing quicker, and offload the long, time-consuming testing to the CIs. So long as the combinations aren't over the top, I don't think that running into build timers would be a big concern.

I'm probably overthinking this though.

**Comment by wlandau-lilly** [comment]
@AlexAxthelm Very clever. It does sound nice to have Travis check all the parallelism configurations automatically, but then again, the added testing time may be a pain for smaller commits.

**Comment by wlandau-lilly** [comment]
With the number of bugfixes that I am preparing, I am also working on a solution to this myself, just to have on hand. Global options seem like another way to handle multiple configurations. 

`R/test.R`:

```r
set_test_opt <- function(opt){
  opt = match.arg(opt, choices = names(test_opts))
  options(drake_test_opt = opt)
}

test_opt <- function(){
  opt <- getOption("drake_test_opt")
  if(!length(opt)) opt <- names(test_opts)[1]
  test_opts[[opt]]
}

test_opts <-
  list(
    parent_parL_2 = list(
      envir = "new.env(parent = globalenv())",
      parallelism = "parLapply",
      jobs = 2,
      cran = TRUE
      ), # For CRAN, Travis, and Appveyor, only use this configuration.
...
```

`tests/multiple.R`

```r
library(magrittr)
library(testthat)
devtools::load_all()

os <- Sys.info()['sysname'] %>% tolower %>% unname

system.time(
  for(opt_name in names(test_opts)){
    set_test_opt(opt_name)
    cat(opt_name, "\n")
    opt <- test_opt()
    if(length(opt$skip_os))
      if(os %in% opt$skip_os){
        cat("  Skipping.\n")
        next
      }
    test_dir("testthat")
  }
)
```

This approach requires running `multiple.R` manually, but for a first pass, that may be okay. It prints the progress nicely, and I am not sure whether I want all these configurations to run under `devtools::check()`.

**Comment by AlexAxthelm** [comment]
`withr` can also change the global options temporarily.

**Comment by wlandau-lilly** [comment]
I should really check that out.
